proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,634 @@
+  scope.
+- If the qualified name has a qualifier, then we work out what
+  the qualifying name refers to. If it refers to a module, then we look at
+  the bindings exported by the module to find one matching the name.
+
+That is, we interpret ``A.b`` by first working out what ``A`` refers to, as a
+module, and then looking inside it to find ``b``. ``A`` may itself be
+qualified (since modules may export module bindings, just like any other
+bindings), hence the need for the recursive lookup.
+
+Exporting module names
+^^^^^^^^^^^^^^^^^^^^^^
+
+Module names *cannot* be exported, since there are no export specifiers that can
+refer to them. [#]_ The existing ``module M`` export specifier
+continues to work as it is defined in the Haskell report.
+
+.. [#] See `Module export specifiers`_ for the changes to allow this.
+
+Imports
+^^^^^^^
+
+Imports at the top level of a file module are treated specially. Their
+target is interpreted as a *dotted* name, which is resolved using the
+current rules for locating external Haskell modules. [#]_
+
+.. [#] For imports in other settings, see `Local modules`_.
+
+Importing a module brings all the exported bindings of the target module
+into scope, as described in [hsmods]_. It also brings
+the target module into scope bound to its declared simple name (which may be
+a dotted name if it is a file module).
+
+If a module is imported ``qualified`` then the target module is no
+brought into scope.
+
+If a module is imported ``as`` a simple name, then the target module is brought
+into scope with that simple name instead.
+
+If multiple modules are imported ``as`` the same name, then we retain the current
+behaviour, which is to “merge” modules. [#]_ The semantics is roughly that:
+
+.. [#] See `What should happen if multiple modules are defined
+       with the same name?`_ for discussion.
+
+::
+
+   import A as M
+   import B as M
+
+means the same as
+
+::
+
+   module M (module A(..), module B(..)) where
+       import A
+       import B
+
+This retains the current behaviour with respect to clashing names, since
+Haskell does not issue ambiguity warnings on import, only on usage. Thus
+the hypothetical definition of ``M`` is error-free even if ``A`` and
+``B`` export ``n``, but an attempt to use ``M.n`` will result in an
+ambiguity error. [#]_
+
+.. [#] This explication of the semantics uses a local module definition,
+   but we could wire the behaviour in as a special case if we decide
+   not to support local modules.
+
+Module export specifiers
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The previous section allows us to refer to modules by name, but not to
+export those names.
+
+We propose adding a language extension ``ExportModuleNames`` which adds the following
+export specifiers:
+
+- ``module M`` *changes* to exports the name ``M`` as a module binding.
+- ``module M(a, b, .. c)`` exports ``M`` and some of its names.
+- ``module M(..)`` exports ``M`` and all of its names. [#]_
```

Comment:
> `M` and all of its names

This doesn't appears to include all modules that `M` exports, right?

Did you consider allowing transparent re-export of modules-exported-by-modules?

Reply:
I'll note that at this point I would switch from being anti-#220 (wart fix is too ad-hoc / bikesheddy) to pro-#220 (imports should match export).

I am likewise sympathetic to the `qualified module` syntax from #205 for matching the status quo in a non-obvious way, but I agree that with the first class modules / "dotted names are bad" approach, the status quo is too harmful to further in enshrine in new features.

Reply:
If you mean:
- `M` has an old-style `module A` export, then that just turns into a bunch of normal exports, so is covered by this.
- `M` has a new-style `module A` export, then that just exports `A` under its name, so that name is covered by this.

Put it this way: I gloss a `M (..)` export as `forall n e . exports(m, n, e), exports(<current module>, n, e)`